### ISSUE TITLE
double.nan has a platform specific bit pattern

### ISSUE THREAD

**Issue by simc** [dart-lang/sdk#49423]
`main.dart:`
```dart
void main() {
  final data = ByteData(4);
  data.setFloat32(0, double.nan, Endian.little);
  print(data.buffer.asUint8List());
}
```

running `dart main.dart` prints:
- `[0, 0, 192, 127]` on macOS arm64
- `[0, 0, 192, 255]` on macOS x64 (using rosetta)

Both are valid `double.nan` representations but I would expect the result to be platform independent. It makes writing tests for a binary format like protobuf much more difficult.

Dart version: 2.17.3

**Comment by lrhn** [comment]
tl;dr: It comes from `divsd 0.0, 0.0` on X64 giving a negative NaN.

The `double.nan` value should preferably be the positive zero-payload QNaN Float64 value.
It seems to be the sign bit which differs here, when it's written as Float32.

I get the same `[0, 0, 192, 255]` result on Linux x64, so it's not just macOS.
The sign bit is also set when doing `setFloat64`, so not just a Float32 issue.

I can't directly see whether the `double.nan` value itself is negative, other than by writing its bits, but this result suggests that it is.
We have one operation which changes the sign even of a NaN, but it doesn't tell us what the original bits were.
```dart
print(identical(double.nan, -double.nan)); // false
```
and if I try to write `-double.nan`, I do get the `127` MSB.

So, either writing the bits of `double.nan` changes the sign bit (would be weird), or `0.0/0.0` is, for some reason, -QNaN instead of QNaN. Let's assume the latter.

Could we define `double.nan` in a different way that `0.0 / 0.0` to get a positive result? 
Let's try the other operations specified to give NaN:
```dart
var zdz = 0.0 / 0.0;                                         
var ifmif = double.infinity - double.infinity;               
var ifdif = double.infinity / double.infinity;           
var ztif = 0.0 * double.infinity;                            
                                                             
print(identical(zdz, ifmif)); // true
print(identical(zdz, ifdif)); // true                      
print(identical(zdz, ztif)); // true                      
```
So, no help, they're all negative. We could do `static const double nan = -(0.0 / 0.0);`, but that would likely give a negative NaN on other platforms which don't have this issue. (And `(0.0 / 0.0).abs()` propagate the NaN value, it doesn't change its sign. Every operation other than unary `-` does.)

Checking the intel `divsd` operation specification doesn't help much. It doesn't specify the behavior. The IEEE-754 (2008) standard allows "some diagnostic information" on the QNaN of `division(0, 0)`, but doesn't say which. So it's not *wrong* for it to be negative, just weird an unexpected.

It does seem that `divsd` on 0.0 and 0.0 yields -QNaN on intel/AMD x64 platforms.
Other [have](https://gamedev.net/forums/topic/661022-why-does-divison-by-0-always-result-in-negative-nan/5180648/) [https://stackoverflow.com/questions/66295777/any-reason-why-0-0-0-0-makes-negative-nan-in-c](noticed) too.
(It's not the compiler, as someone suggests, it's just that clang does better inlining and decides that 0.0/0.0 is +QNaN at compile-time, but if it gets to runtime and executes divsd, it gets -QNaN too.)

See also this [compiler comparison](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIMwCcpK4AMngMmAByPgBGmMQSGqQADqgKhE4MHt6%2B/kGp6Y4CYRHRLHEJXEl2mA6ZQgRMxATZPn6Btpj2RQwNTQQlUbHxibaNza25HQrjA%2BFD5SNVAJS2qF7EyOwc5gDM4cjeWADUJrtuTjPEmKxn2CYaAIJe6UbHzGwKyUxbxzPoZysTwej3Q6xi9GOLAAnvgAG4QOGeJiOSFgrwQzDHJikY5IgyorHozHHGLLU4AdiBj2OtOO1wIGwY2OOAHpSYCQSYKQARLlPcIEKFMcIQcnc6l0vHIwnHYmQuFiLxYs48qGwvAIjQAOiSxx1GmWnKeUq8DHSwAi6GOtAEwBtdrxp12aoAVNdBfFkgyAPqiGZnNz4lF0LFmi1Wh1vW1GV13CDmABsiu8mCNu0ldLQXgIgcDxwQmFUzrc%2BbhefOrloxppdIZTP1Ne5PI4q1onAArLw/BwtKRUJxS5ZrH91psVWZdjxSARNK3VgBrEC7Xa6yQU3YaTeSAIADgpgQp%2Bk4km7c/7nF4ChASVnvdbpDgsBgiBQqBYyVDZAoEDQH6/KAGEYXBVFwfB0AQ8TXhAMTnjE4RNNCnDTn%2BbCCAA8gwtBIfepBYCwhjAOIuH4NcdRwpg164UWtQ5tsfaeu2uG0HgMTEIhHhYMhvAEMQeAsNxqxUAYwAKAAangmAAO7ockjDcTIggiGI7BSIp8hKGo566GBQHGNY1j6Kx16QKsqDJD0VEALQAOpiLQxy2UWvFMI5cnoIYjjII5/yqqYw6WFwFKOehuxXl0tQ9C4DDuJ4bR6KE8xlBUegFBkAiTH4YFpT0gzJUsnTdPUsyZXoNR1AIfTNHlwyVGM/SlWBMz9DViyVKsChjlsei8Zg2w8G2nZnrhA4cKou6JlZiaSMchyEccIG6tqXDHBAQ5WJYuK4IQJCnJOYHHB4/70MQe1TssPFzssqyFkwWAJGKx4cKepACR2SQ9n2o1XjeM5XY%2BL4QEg6wEMkObkC4%2BBEHx6B6PwSmiOIanwxpKjqLhOmkFJ7HJIJT1dqQn28KN6E5mDQqoFQxzjZN02zXpC1cEtK1re%2Bn4nWdXAXX997XaQS7vU9L1vR957fbYv13lofNLoE2q7hoGhcLugSSFuFKJhuiZPWFhNi5ePPSwD8BPq%2Bf7s/EEO/mzAHAFUZjgbQkHENBsG4fBzDEDhKHvmhBCYdh574YRxF9qRkWapR540cgdHnox54sWxHEYP1PF8QJ3APsJTCiRJ0myfJWfqcpSPSCjiho9p/j6IRKAGZtScmY95mWZwtn2Y5NnOexbmYB5gh4N5Vm%2BS6/kbRYQUhbr5VRRAriNcEMWtSl2VpOlWRxbka%2BFJkK8FbPxUNVvWWFRHlWzPvdXNRMJ9lZfSW1RIHVdapM7XGng0cATRMXmNE1TRmsAZA3kqjajMKtdahljjbWhpzXER0LanT2GYbmUt5z82XKuFcODcF4KFrwEWesRoG2vLef6psgYgBBuTK2/cdowzhrIUuqly6yErlpDGNdsZMFxlnL%2BP99YcFJqDHMxxKbUwAXTYBoCdQQNZsdeIZ1UGXV5jdG491KBf2FiAQWv9xZkMNhgpcuwKTLQ7GYDsUhFYBApO9XYOthpfQNugvmTEzBOOJi4q6qwKIu0yCASQQA%3D%3D%3D).

So, what can we do?

Maybe we can make `nan` an "external constant" and let the compiler provide the value as a `reinterpret_cast<double>(0x7FF8000000000000ull)`.

Or just accept that NaNs are weird and live with it. I'd prefer to not do that. At least, if we can provided a non-negative zero-payload QNaN as `double.nan`, we can better live with `0.0/0.0` giving a negative QNaN, because we can get to the non-negative value somehow.

**Comment by mraleph** [comment]
It's a known ARM vs X86 difference. If an operation would produce NaN and has no NaN as an input X86 will produce 1 in the sign bit and ARM produces 0.

**Comment by lrhn** [comment]
Since any user created `0.0/0.0` will give the same signed NaN, we can't prevent the signed NaN from showing up. 
On second thought, it might actually be safer to have `double.nan` be identical to the NaNs you get from standard NaN-generating operations.

So, leaning towards doing nothing here.

**Comment by mraleph** [comment]
> So, leaning towards doing nothing here. 

Should we close this then? I am also in the do nothing camp for this one. (Maybe add a documentation line into the `double.nan` that is has a platform specific bit pattern).

**Comment by leisim** [comment]
I think at least documenting this behavior is a good idea. Here is a quote from the Rust docs:

> Note that IEEE-745 doesn’t define just a single NaN value; a plethora of bit patterns are considered to be NaN. Furthermore, the standard makes a difference between a “signaling” and a “quiet” NaN, and allows inspecting its “payload” (the unspecified bits in the bit pattern). This constant isn’t guaranteed to equal to any specific NaN bitpattern, and the stability of its representation over Rust versions and target platforms isn’t guaranteed.

Shall I open a PR with a doc line?

**Comment by lrhn** [comment]
I'd add documentation to `double.nan` for all things related to NaN.

All other places that need to say something about NaN should then refer to that.
Not sure `setFloat32`/`setFloat64` need to say anything, they are dutifully storing the correct bit pattern.

We could mention on `/`, `-` and `*` that a NaN result from non-NaN inputs is not guaranteed to be any specific pattern,
or just say that it'll be the same as `double.nan` (which I think it is on all platforms).
